lib/fsm/utils.go

```diff
@@ -43,9 +43,85 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
 		return trace.BadParameter(
 			"phase %q has already been rolled back", phase.ID)
 	}
+
+	// TODO: Rollback of non-leaf phases is not currently supported.
+	// Rollback starts top-down, and not in reverse order.
+	if phase.HasSubphases() {
+		return trace.BadParameter(
+			"unable to rollback non-leaf phase, only leaf phases can be rolled back").AddField("phase", phase.ID)
+	}
+
+	if !dependentsRolledBack(plan, phase.ID) {
+		return trace.BadParameter(
+			"rollback subsequent phases before rolling back phase %q", phase.ID)
+	}
 	return nil
 }
 
+// dependentsRolledBack returns true if all phases that are dependent on the
+// phase specified by phaseID is unstarted or has been rolled back.
+func dependentsRolledBack(plan *storage.OperationPlan, phaseID string) bool {
+	// required will be nil if an invalid phaseID is provided.
+	required := getRequired(plan.Phases, phaseID)
+	if required == nil {
+		return true
+	}
+	return dependentsRolledBackHelper(required, plan.Phases)
+}
+
+// getRequired constructs the initial set of required phases. This set includes
+// the phase specified by phaseID and its parent phases.
+func getRequired(phases []storage.OperationPhase, phaseID string) map[string]struct{} {
+	for _, phase := range phases {
+		if phase.ID == phaseID {
+			return map[string]struct{}{
+				phaseID: {},
+			}
+		}
+		required := getRequired(phase.Phases, phaseID)
+		if required != nil {
+			required[phase.ID] = struct{}{}
+			return required
+		}
+	}
+	return nil
+}
+
+// dependentsRolledBackHelper is a recursive helper function that returns true
+// if all phases that are dependent on the required set is unstarted or has
+// been rolled back.
+func dependentsRolledBackHelper(required map[string]struct{}, phases []storage.OperationPhase) bool {
+	if len(phases) == 0 {
+		return true
+	}
+
+	for _, phase := range phases {
+		if isDependent(required, phase) {
+			if !phase.IsUnstarted() && !phase.IsRolledBack() {
+				return false
+			}
+			required[phase.ID] = struct{}{}
```

Comment:
Hmm, do we need to add this phase to the "required" set as well? I wonder if it'll be simpler to just leave the original set as it is, this way the algo will be just one dependency "deep", right? So for example, in case of the following plan:

```
/init
/masters <--depends on /init
/gc <-- depends on /masters
```

when you try to rollback /init and /masters and /gc are in progress, it would only tell you to rollback /masters first. Then you try to rollback /masters and it will tell you to rollback /gc first.

We could probably even get rid of recursion in this case, just iterate over all phases and see if any of them depend on those in the "required" set.

What do you think?

Reply:
That approach would definitely be a lot simpler, but I'm not sure if the check will function properly if a phase is executed out of band. Since phases can be executed out of order, wouldn't something like this be a valid state of the plan?
```
/Init { state: completed }
/masters { state: unstarted, requires: { /init } }
/gc { state: completed, requires: { /masters } }
```
In this situation all phases that require /init are either un-started or rolled back.
Would /gc in this case even have been able to be executed?

Reply:
Yes, you're right. Let's keep it as-is then, but add a comment here (as it's also a side-effect to the passed in parameter which is not great) and a test case for this scenario.